crates/shared/src/bad_token/token_owner_finder/ethplorer.rs

```diff
@@ -0,0 +1,167 @@
+            rate_limiter: None,
+        })
+    }
+
+    pub fn with_rate_limiter(&mut self, strategy: RateLimitingStrategy) -> &mut Self {
+        self.rate_limiter = Some(RateLimiter::from_strategy(strategy, "ethplorer".to_owned()));
+        self
+    }
+
+    async fn query_owners(&self, token: H160) -> Result<Vec<H160>> {
+        let mut url = self.base.join(&format!("{token:?}"))?;
+        // We technically only need one candidate, returning the top 2 in case there
+        // is a race condition and tokens have just been transferred out.
+        url.query_pairs_mut().append_pair("limit", "2");
+
+        tracing::debug!(%url, "querying Ethplorer");
+        // Don't log the API key!
+        url.query_pairs_mut().append_pair("apiKey", &self.api_key);
+
+        let request = self.client.get(url).send();
+        let response = match &self.rate_limiter {
+            Some(limiter) => limiter.execute(request, back_off::on_http_429).await??,
+            _ => request.await?,
+        };
+
+        let status = response.status();
+        let status_result = response.error_for_status_ref().map(|_| ());
+        let body = response.text().await?;
+
+        tracing::debug!(%status, %body, "response from Ethplorer API");
+
+        // We need some special handling for "not a token contract" errors. In
+        // this case, we just want to return an empty token holder list to conform
+        // to the expectations of the `TokenHolderProposing` trait.
+        if status == StatusCode::BAD_REQUEST {
+            let err = serde_json::from_str::<Error>(&body)?;
+            if err.not_token_contract() {
+                return Ok(Default::default());
+            }
+        }
+        status_result?;
+
+        let parsed = serde_json::from_str::<Response>(&body)?;
+
+        Ok(parsed
+            .holders
+            .into_iter()
+            .map(|holder| holder.address)
+            .collect())
+    }
+}
+
+#[derive(Deserialize)]
+struct Response {
+    holders: Vec<Holder>,
+}
+
+#[derive(Deserialize)]
+struct Holder {
+    address: H160,
+}
+
+#[derive(Deserialize)]
+struct Error {
+    error: ErrorData,
+}
+
+#[derive(Deserialize)]
+struct ErrorData {
+    code: i64,
+}
+
+impl Error {
+    fn not_token_contract(&self) -> bool {
+        // https://github.com/EverexIO/Ethplorer/wiki/Ethplorer-API#error-codes
+        self.error.code == 150
+    }
+}
+
+#[derive(MetricStorage, Clone, Debug)]
+#[metric(subsystem = "ethplorer_token_owner_finding")]
+struct Metrics {
+    /// Tracks number of "ok" or "err" responses from blockscout.
+    #[metric(labels("result"))]
+    results: IntCounterVec,
+}
+
+#[async_trait::async_trait]
+impl TokenOwnerProposing for EthplorerTokenOwnerFinder {
+    async fn find_candidate_owners(&self, token: H160) -> Result<Vec<H160>> {
+        let metric = &Metrics::instance(global_metrics::get_metric_storage_registry())
+            .unwrap()
+            .results;
```

Comment:
When Valentin refactored metrics code the last time he stored a `&'static Metrics` in the struct instead of calling `Metrics::instance()` every time which I found quite elegant.
This would then also remove 1 `unwrap()`.

Reply:
Nice idea. Implemented.